### ISSUE TITLE
Documentation: Sequential.fit documentation - missing information on class_weight

### ISSUE THREAD

**Issue by mjs-wpi** [fchollet/keras#5116]
class_weight = 'auto' is not documented online or in help(Sequential.fit)

**Comment by bstriner** [comment]
@mjs-wpi What is `class_weight='auto'` and how does it differ from `class_weight='None'`?

Cheers

**Comment by baziotis** [comment]
@bstriner  I think he has in mind something like the `class_weight='balanced'` that many classifiers in scikit-learn have. What it does is that it automatically finds the weights for each class (for imbalanced datasets).

@mjs-wpi  In keras you have to pass the weights on you own.
look at this https://github.com/fchollet/keras/issues/1875

**Comment by bstriner** [comment]
So, is this a feature request to add something like `class_weight='balanced'` to Keras? It would just count each class and create balanced weights like your code snippet. Seems like a reasonable PR.

The phrasing of the title makes it sound like something that is implemented but not documented.

**Comment by baziotis** [comment]
@bstriner  i think @mjs-wpi assumed that this is a functionality Keras already has, but is not documented. To be fair it would be nice if we had something like this. Since we are talking about this i want add a note.

Recently i worked with a very imbalanced data set (classA:25%, classB:74%, classC:1%). if we use the weights produced by simple their percentages then classC will get a very big weight and as a result the model will end up biased towards C a lot more than it should. So i added i smoothing factor that you can tweak. It worked better for me. `smooth_factor==0` means no smoothing.
```python
def get_class_weights(y, smooth_factor=0):
    """
    Returns the weights for each class based on the frequencies of the samples
    :param smooth_factor: factor that smooths extremely uneven weights
    :param y: list of true labels (the labels must be hashable)
    :return: dictionary with the weight for each class
    """
    counter = Counter(y)

    if smooth_factor > 0:
        p = max(counter.values()) * smooth_factor
        for k in counter.keys():
            counter[k] += p

    majority = max(counter.values())

    return {cls: float(majority / count) for cls, count in counter.items()}
```

**Comment by mjs-wpi** [comment]
Hi everyone,

I think I found the source of the problem. I was calculating my own class weights (very unbalanced classes) and was getting nonsensical loss function results for a 6-class softmaxed final layer: I expected around 1.7 and was getting 0.08. I thought this may have to do with my weight calculations. When I instead tried class_weight = 'auto' per a misguided stackoverflow comment, it seemed to work.

I then assumed that is must have been an undocumented feature in Keras. After reading your comments, I tried running my code with class_weight = 'nonsensical_word' and it also gave me good results. 

So because an error message wasn't thrown and things got better, I assumed Keras had an undocumented feature.

@baziotis Thank you for posting the smoothing function. I've been having a lot of trouble debugging why my loss function is so erratic and I think smoothing may help a lot.

**Comment by baziotis** [comment]
@mjs-wpi FYI i have found that values around 0.1 are a good default for very _imbalanced_ classes.

**Comment by mjs-wpi** [comment]
Thanks @baziotis, I will give it a try!

**Comment by bstriner** [comment]
Stackoverflow is hit or miss. Maybe the code was interpreting the string as a list of weights or something. It probably should have been throwing an error.

**Comment by mhirimohamed** [comment]
how can I use class_weight with binary_crossentropy loss ?

**Comment by alyato** [comment]
@cbaziotis 

> param y: list of true labels (the labels must be hashable)

my Y_label is `(number,)`,how to rewrite it? 
And `counter = Counter(y)`,how to define the function of `Counter()`? Thanks

**Comment by aghoshpub** [comment]
@alyato:
`from collections import Counter`

@cbaziotis

it should be:
 `return {cls: float(majority) / count for cls, count in counter.items()}`
